Attach address to shipment

reinaris opened the pull request:
Currently a shipment address is stored on the order.
I'm working on this feature where you can change the address per shipment.

The address is still stored on the order by default, but now you can supply a new address to a single shipment. Very useful if you have multiple shipments.

The best way to explain is with screenshots:
## Change address feature on the order/shipment page

![screen_shot_2015-02-04_at_09_54_34](https://cloud.githubusercontent.com/assets/106335/6038408/addcca0a-ac54-11e4-8648-7c0a78541c14.png)
## Supply a new address modal

![screen_shot_2015-02-04_at_09_55_23](https://cloud.githubusercontent.com/assets/106335/6038412/b1da9524-ac54-11e4-9edf-0e9052594aec.png)
## A different address for each shipment

![screen_shot_2015-02-04_at_09_55_50](https://cloud.githubusercontent.com/assets/106335/6038419/c3adf1d8-ac54-11e4-9a2b-6617c6da5cad.png)

It still needs:
1) Specs

While working on this feature a couple of things came up to discuss:
1) Why not store the address to a shipment by default
2) In this version (0.1) I only made it possible to add a new address to a shipment. But it should be nicer to also be able to change the current address. I din't took this approach because it is not clear that you're editing the order address (if you didn't created a new address for a shipment first).
3) In general I think we need to think about the order.ship_address comparing to saving the address to a shipment.

I think even this current PR is very usable, but it needs some discussion for the future.

Let me know what you think about this!

reinaris committed:
Be able to change the address of a specific shipment

reinaris committed:
Added some comments to country/states javascript

reinaris committed:
Improved some html for shipment addresses

reinaris committed:
Renamed change to 'new shipment address'

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://reinaris-attach-address-to-shipment.spree-spree-9ed724b77adec04c49c2.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://reinaris-attach-address-to-shipment.spree-spree-9ed724b77adec04c49c2.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

BenMorganIO commented:
:+1:

reinaris committed:
Improved js quality country_states

reinaris committed:
Improved javascript orders/edit and country_states

JDutil commented:
@reinaris we definitely want to move to associating the shipping addresses directly to each shipment rather than order, and already made the association available.  It's mostly a matter of just making sure we don't break backwards compatibility without deprecating it first.  Would be great to eventually allow multiple shipping addresses to be entered on the frontend as well.

BenMorganIO commented on backend/config/routes.rb:
```diff
@@ -1,5 +1,7 @@
 Spree::Core::Engine.add_routes do
   namespace :admin do
+    post '/addresses/create_shipment_address/:shipment_id/', to: "addresses#create_shipment_address", as: :create_shipment_address
```
This doesn't look very RESTful. Its assumed that a create occurs because its a POST and I don't think address is needed at the end since `/addresses` is already specified. Maybe:

```ruby
post '/addresses/shipment/:shipment_id', to: "addresses#create_shipment", as: :create_shipment_address
```

reinaris replied:
Agree! Thanks for your feedback!

BenMorganIO replied:
I think we're in the 100 character limit now :D

reinaris committed:
Improved address create_shipment method name and routes

BenMorganIO commented on backend/config/routes.rb:
```diff
@@ -1,5 +1,7 @@
 Spree::Core::Engine.add_routes do
   namespace :admin do
+    post '/addresses/shipment/:shipment_id', to: "addresses#create_shipment", as: :create_shipment_address
```
Nope, take tat back, still 6 over...

reinaris replied:
`as: :shipment_address` ? :)

BenMorganIO replied:
lol I think that'll work and follow rails conventions! :D

BenMorganIO replied:
I wonder what the named route would be if you dropped the `as` too.

JDutil replied:
I think I'd rather have the controller action a member of the shipment controller rather than the address controller.

simmsy commented:
@JDutil I don't think we would want to deprecate having a default  order shipping address as this 'should' arguably be used when creating new shipments. And yes I agree, setting different addresses during the checkout would be great, and I believe this PR would facilitate that when processing the shipments further downstream.

I believe is @reinaris refactors the can_get_rates? method to use his new shipment_address method then this allows Spree to support shipment addresses (only if present).

Also, if @reinaris only allows the ability to edit addresses attached to a shipment on the shipment screen (or add one if shipment is using the order's ship address) then this should avoid any confusion over what address is being edited on this screen. The user can still edit the master order ship address on the main order edit page.

It seems that Spree was intended to set the shipment's address here, but this method must not be used:
https://github.com/spree/spree/blob/master/core/app/models/spree/stock/coordinator.rb#L11:L15

I believe with these changes this PR will enable supporting shipment address without breaking BC.

reinaris committed:
can_get_rates uses new shipment_address method

dhonig commented:
+1 for  shipment-shipping_address to be a 1..1 relationship.

reinaris committed:
Be able to apply the new address to other shipments

Did some code formatting and improvements
Deleted unused javascript
Cleaned up shipment address form

reinaris committed:
Improved orders/customer page

Display a preview of the address
Edit addresses in modals
Better layout of the address form

reinaris committed:
Added default state_name to form

reinaris committed:
Use the address of the original shipment on transfer

When you transfer an item to a new shipment we now take over the
address of the original shipment.

reinaris committed:
Show the normal form without modal while adding an order manually

And did some translation of the new alert message

reinaris committed:
Improved customer ship address warning message

JDutil commented:
Orders should have a default shipping address I just want to make sure we're not breaking the existing API without proper deprecations where needed.  That really isn't an issue until dealing with the frontend & new backend order form though.

JDutil commented on core/app/models/spree/shipment.rb:
```diff
@@ -262,6 +262,10 @@ def shipping_method
     end
 
+    def shipment_address
+      self.try(:address) || order.ship_address
```
Why create a new method here?  I think a shipment should always reference it's own address.

simmsy replied:
@JDutil There is a BC issue in that when splitting to a new shipment it doesn't set a delivery address on the new shipment. @reinaris has fixed this but for BC reasons we should assume you can have a nil address and therefore must inherit the order's shipping address. An alternative (and maybe better approach) is to add a migration to set any missing addresses to the default shipping address.

Also, @reinaris will be submitting some changes to this PR soon that effectively make addresses immutable from the admin (by editing you create a new one). This will ensure data sanity when editing addresses that were originally shared with multiple shipments (without cloning them originally). We have also changed the edit address in the admin a little to make this easier to understand.

BenMorganIO replied:
By BC do  you mean British Columbia :D

JDutil replied:
IMHO I'd prefer we did a migration to cleanup nil address data issues rather than work around them.

reinaris replied:
Agree with that, much cleaner. Will work on it!

TrangPham replied:
@BenMorganIO BC means backwards compatibility here.

I agree with @simmsy for writing a migration

BenMorganIO replied:
@TrangPham Thankyou lol. Now everything makes sense.

JDutil commented on backend/app/assets/javascripts/spree/backend/admin.js.erb:
```diff
@@ -7,6 +7,7 @@
 //= require spree/backend/product_picker
 //= require spree/backend/option_value_picker
 //= require spree/backend/taxons
+//= require spree/backend/country_states
```
If this country_states file isn't the same as the frontend's code for switching country/states (which I don't think it is) we should probably update the frontend to use it as well. I'd like to keep it DRY, which not sure how to best do keeping frontend/backend separate.  I don't think core should concern itself with this, but perhaps we should keep it in API?

reinaris replied:
True, but the frontend javascript was not suitable for the backend so I had to clone and edit this for the backend.
However there are now 2 country/states javascripts in the backend. Which the one I created has some more functionality, so maybe we should deprecate the old one.

reinaris replied:
I removed the old update_state method from the backend and replaced all with the new javascript.
The old method did not checked if the country needed a state, so a text field was always shown. This method does.

reinaris committed:
Replaced address_states javascript with newer country_states

reinaris committed:
Improved code quality states javascript and addresses controller

reinaris committed:
Made shipment address uneditable after order complete

Added message to explain that addresses should be edited on the shipment

reinaris committed:
Cleanup shipment address change modal

You can only change a shipment address, so I deleted the type
(update/new)

reinaris committed:
Deleted the shipments helper

This was a helper created during developing, not needed at the end

reinaris committed:
Mixed double and single quotes.

reinaris commented:
I did more work to make this user friendly and solid.
Some screens!

## Order customer details page
![screen_shot_2015-02-10_at_11_59_54](https://cloud.githubusercontent.com/assets/106335/6126909/1e2b5bf6-b11d-11e4-8e95-ff323a7e692f.png)

## Order shipments page
![screen_shot_2015-02-10_at_12_00_15](https://cloud.githubusercontent.com/assets/106335/6126910/21527148-b11d-11e4-9b23-a6733cd1ddc6.png)

## Change a shipment address
![screen_shot_2015-02-10_at_12_00_44](https://cloud.githubusercontent.com/assets/106335/6126913/2fdf3e30-b11d-11e4-9db8-077447bc4f62.png)

In my opinion this works perfect if you need to edit shipment addresses. The address is in the same place as the other shipment info (tracking number, shipping method etc.). Not at the customer page anymore what makes sense to me.

The feature to apply a changed shipment address to multiple shipments makes it easy to update multiple shipments.

A shipment address is never updated but a new one is created, so it's immutable.
We took this approach so you will never create a situation where a shipment address is changed and an other shipment address changed as well while the user didn't noticed.

Of course this needs some specs but I first like to hear your opinion.
From a UI view this is a nice improvement.

peterberkenbosch commented:
:+1: from me! @reinaris do we keep track of older shipment address? Could be an option to see the changelog of the changes for CS.

reinaris commented:
Thanks!
We currently not track older shipment addresses, but that could be an improvement for the future.

JDutil commented on backend/app/views/spree/admin/orders/_shipment_address_form.html.erb:
```diff
@@ -0,0 +1,111 @@
+    </p>
+  <% end %>
+
+  <% other_shipments = @order.shipments.where.not(id: shipment.id, state: 'shipped') %>
```
@simmsy mentioned desire to add new shipment states as well soon so we may want to make this state check explicitly for pending or ready rather than not shipped.

reinaris committed:
Added scope for editable shipment. Added migration for shipment address

Added a scope editable to the shipments which we use in the shipment
address modal.
Added a migration so we dont have to introduce a new method
(shipment_address).

reinaris commented:
In my last commit I added a migration instead of introducing new methods like @JDutil suggested. I also added a scope for editable shipments.

Please let me know what you think of this PR so I can start finishing it up with some specs :)

simmsy commented:
:+1: from me!

athal7 commented on backend/app/assets/javascripts/spree/backend/country_states.js:
```diff
@@ -0,0 +1,85 @@
+  });
+});
+
+function requestStates($parent){
```
would it make sense to namespace these functions rather than putting them on the window?

BenMorganIO replied:
@athal7 :+1:

JDutil replied:
:+1:

reinaris replied:
Agree! Will work on that tomorrow!

athal7 commented on core/app/models/spree/shipment.rb:
```diff
@@ -36,6 +36,7 @@ def generate_number(options = {})
     scope :ready,   -> { with_state('ready') }
     scope :shipped, -> { with_state('shipped') }
     scope :trackable, -> { where("tracking IS NOT NULL AND tracking != ''") }
+    scope :editable, -> { where(state: ["pending", "ready"]) }
```
to me it doesn't seem that ready shipments should be editable. ready seems to convey that it is ready to ship such that a 3rd party should expect that it will not change and can safely ship this package.

simmsy replied:
@athal7 I agree but I think this is a bigger discussion. For now at least this scope follows Spree's current behaviour and provides a hook that you can override for finding editable shipments. We will follow this PR soon with another one that deals with this issue separately. That cool?

athal7 commented on core/app/models/spree/shipment.rb:
```diff
@@ -344,7 +345,9 @@ def transfer_to_location(variant, quantity, stock_location)
+          stock_location: stock_location,
+          address: self.address)
 
         order.contents.remove(variant, quantity, {shipment: self})
```
this isn't your code, but it seems odd that we are doing order contents things here, maybe just grabbing quantity # of inventory units and moving them over?

athal7 commented on core/app/models/spree/shipment.rb:
```diff
@@ -381,7 +384,7 @@ def after_ship
 
       def can_get_rates?
-        order.ship_address && order.ship_address.valid?
+        address && address.valid?
```
maybe check order.ship_address if there is no address?

simmsy commented on core/db/migrate/20150213094131_attach_shipments_to_order_address.rb:
```diff
@@ -0,0 +1,10 @@
+class AttachShipmentsToOrderAddress < ActiveRecord::Migration
+  def up
+    Spree::Shipment.where(address: nil).each do |shipment|
```
It would be play nicer for larger stores if we used update_all here, so something like:

`Shipment.where(address:nil).joins(:order).update_all("shipments.address_id = orders.ship_address_id")`

JDutil replied:
:+1:

reinaris replied:
Agree! I was already looking for something faster but couldn't come up with you solution :) will improve this!

reinaris committed:
Namespaced all the country_states javascript functions

reinaris committed:
Fallback to order ship_adress if the shipment has no address

There can be situation where an shipment has no address. If you create
it manually or maybe trough the API.

reinaris committed:
Corrected can_get_rates method, check for an addres first

reinaris committed:
Corrected can_get_rates method, forgot a question mark..

krtschmr commented:
great work!

bazfer commented:
What's the status of this branch?

reinaris commented:
Well it's pretty old :)
I'm not on to this anymore, sorry. Anyone who would like to pick it up; go ahead!

damianlegawiec commented:
Thanks @reinaris for all your contributions to Spree 👍 We wouldn't be here without your work :) As of this PR - we're going to include parts of it in [future Spree versions](https://github.com/spree/spree/milestones?direction=asc&sort=due_date&state=open) alongside [merging Spree Address Book into core](https://github.com/spree/spree/issues/6942)

bazfer commented:
Appreciate your responses!

On Tue, Sep 26, 2017 at 3:25 AM, Damian Legawiec <notifications@github.com>
wrote:

> Thanks @reinaris <https://github.com/reinaris> for all your contributions
> to Spree 👍 We wouldn't be here without your work :) As of this PR -
> we're going to include parts of it in future Spree versions
> <https://github.com/spree/spree/milestones?direction=asc&sort=due_date&state=open>
> alongside merging Spree Address Book into core
> <https://github.com/spree/spree/issues/6942>
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/spree/spree/pull/6001#issuecomment-332124698>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ATqdi7L7qO4stRzlA-Kyuks6MnzH-MK9ks5smLUQgaJpZM4Dbffc>
> .
>



-- 

*Fernando Baz*
*Software Developer*

bbonislawski commented:
This was rebased in #7350 . Closing this in favor of rebased version
